Sources/Linq2DynamoDb.DataContext/TableDefinitionWrapper.cs

```diff
@@ -360,45 +369,120 @@ protected override void InitReader(ISupervisableEnumerable reader)
 
-            // this should never throw any exceptions
-            this.Cache.UpdateCacheAndIndexes(addedEntities, modifiedEntities, removedEntities);
-
-            return dynamoDbUpdateTask.ContinueWith
+            return Task.WhenAll(dynamoDbUpdateTasks).ContinueWith
             (
                 updateTask =>
                 {
                     if (updateTask.Exception != null)
                     {
                         // If the update failed - then removing all traces of modified entities from cache,
                         // as the update might partially succeed (in that case some of entities in cache might become stale)
-                        this.Cache.RemoveEntities(addedEntities.Keys.Union(modifiedEntities.Keys).Union(removedEntities));
+                        this.Cache.RemoveEntities(addedEntities.Keys.Union(modifiedEntities.Keys).Union(removedEntities.Keys));
 
                         throw updateTask.Exception;
                     }
+
+                    if (!string.IsNullOrEmpty(VersionPropertyName))
+                    {
+                        // Update the cache if there is a Version Property. Version properties change 
+                        // during insertion/updating so we want to have the latest values for versions 
+                        // of added and modified entities. 
+                        this.Cache.UpdateCacheAndIndexes(addedEntities, modifiedEntities, removedEntities.Keys);
+                    }
                 }
             );
         }
 
+        /// <summary>
+        /// Performs create/update operations with DynamoDB, respecting the [DynamoDBVersion] attribute if present
+        /// </summary>
+        /// <remarks>
+        /// If the entity is new (Version property has not been set) set the version property 0. If 
+        /// updating an existing entity increment the Version property by 1.
+        /// </remarks>
+        /// <param name="entitiesToAdd">Dictionary of entity keys and documents to add/update</param>
+        /// <returns>An IEnumerable containing one task for each entity to add or update.</returns>
+        private IEnumerable<Task> AddEntriesAsync(IDictionary<EntityKey, Document> entitiesToAdd) 
+        {
+            return entitiesToAdd
+                .Values
+                .Select(entry => {
+                    var putConfiguration = new PutItemOperationConfig();
+                    if (!string.IsNullOrEmpty(VersionPropertyName)) 
+                    {
+                        if (entry[VersionPropertyName] != null && 
+                            entry[VersionPropertyName].ToPrimitive(typeof(object)).Value == null)
+                        {
+                            entry[VersionPropertyName] = 0;
+
+                            putConfiguration.ConditionalExpression = new Expression() 
+                            {
+                                ExpressionStatement = $"attribute_not_exists(#key)",
```

Comment:
is this not automatically verified internally by dynamo? what would happen if you removed this check and tried to submit 2 entries with same hash key property?